File: primalTestDotNet/AI/States/LeavingLevelState.cs
using PrimalTestDotNet.Entities;

namespace PrimalTestDotNet.AI.States;

public class LeavingLevelState(AIStateMachine context, Level.Level level) : StateBase(level), IState
{
    public void Update(Hero hero, IEnumerable<IGameObject> gameObjects)
    {
        var distanceData = CalculateDistanceData(hero.Position, []);
        // Try to exit the level
        var exit = gameObjects.First(o => o is Exit);
        var pathToExit = GetPath(hero.Position, exit.Position, distanceData);
        var node = pathToExit.Last();
        var direction = node - hero.Position;
        if (direction.Length > 1)
        {
            // A hero only can move one cell. If the path has longer movement vector it must be the end-goal without any real paths to it
            context.ChangeState(context.UnwinnableState);
            return;
        }

        hero.Move(direction);
    }
}

File: PrimalTestCpp/LeavingLevelState.cpp
#include "LeavingLevelState.h"
#include "Hero.h"
#include "Exit.h"
#include "AIStateMachine.h"

LeavingLevelState::LeavingLevelState(AIStateMachine& context, Level& level) 
	: StateBase(level), _context(context) { }

void LeavingLevelState::update(Hero& hero, std::vector<std::shared_ptr<IGameObject>>& gameObjects)
{
	auto distanceData = calculateDistanceData(hero.getPosition(), {});
	// only one exit is supported at the moment
	auto exits = findAllObjectsOfType<Exit>(gameObjects);
	auto& exit = exits.front();
	auto pathToExit = getPath(hero.getPosition(), exit->getPosition(), distanceData);
	auto& node = pathToExit.back();
	auto direction = node - hero.getPosition();
	if (direction.getLength() > 1) {
		_context.changeState(_context.UNWINNABLE_STATE());
		return;
	}

	hero.move(direction);
}
